### PR TITLE
Dashboards: Costs Table

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/1049

## :v: What does this PR do?

Create a new section for the dashboard with the cost table. This table contains two sublevels and the item of each of the costs.

### After this PR

dashboard_costes 2020-06-03 13_21_57

### CODE REVIEW

File: app/views/gobierto_dashboards/dashboards/costs.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      <p>
+        <%= t(".description", entity_name: current_site.name) %>
+      </p>
+    </div>
+  </div>
+
+  <div id="gobierto-dashboards-costs-app"
+       data-site-name="<%= @site.name %>"
+       data-logo-url="<%= @site.configuration.logo_with_fallback %>"
+       data-home-url="<%= gobierto_dashboards_contracts_summary_url %>"
```
Comment (ferblape) [inline app/views/gobierto_dashboards/dashboards/costs.html.erb]: This URL should be gobierto_dashboards_costs_summary_url

File: config/locales/gobierto_dashboards/views/ca.yml [hunk 2]
```diff
@@ -45,6 +45,32 @@ ca:
...
+          income_cost: Ingressos - Cost
+          legal_framework: Marc legal
+          personal_cost: Cost personal
+          public_tax: Taxa o preus públics
+          subsidies: Subvencions
+          total: Total
+          transference: Transferències
+          types: Tipus
+        see_all: Vegeu-ho tot
+        total: Coste total
```
Comment (ferblape) [inline config/locales/gobierto_dashboards/views/ca.yml]: Cost total

File: config/locales/gobierto_dashboards/views/ca.yml [hunk 3]
```diff
@@ -80,6 +106,7 @@ ca:
       application:
         gobierto_dashboards:
           contracts: Contractes i licitacions
+          costs: Costes
```
Comment (ferblape) [inline config/locales/gobierto_dashboards/views/ca.yml]: Costos

**Review by ferblape — COMMENTED** [review]
This looks great ;)

File: app/assets/stylesheets/module-dashboards.scss [hunk 4]
```diff
@@ -237,3 +242,354 @@
...
+
+  &-table-item-left-container {
+    margin-bottom: 1.5rem;
+  }
+
+  &-table-item-title,
+  &-table-item-text {
+    font-size: 1rem;
+    font-weight: normal;
+    color: #666;
```
Comment (Crashillo) [inline app/assets/stylesheets/module-dashboards.scss]: Static colors?

File: app/assets/stylesheets/module-dashboards.scss [hunk 5]
```diff
@@ -237,3 +242,354 @@
...
+    width: 100%;
+    margin-bottom: 3rem;
+    display: flex;
+    flex-wrap: wrap;
+    box-sizing: border-box;
+
+    &-income,
+    &-coverage {
+      .gobierto-dashboards-table-item-right-table {
+        border-color: rgba(33, 150, 83, .2);
```
Comment (Crashillo) [inline app/assets/stylesheets/module-dashboards.scss]: Static colors?

File: app/assets/stylesheets/module-dashboards.scss [hunk 6]
```diff
@@ -237,3 +242,354 @@
...
+
+  &-table-item-right-section {
+    width: 22%;
+    display: inline-block;
+    padding-right: .5rem;
+  }
+
+  &-table-item-right-section-title {
+    font-size: 1rem;
+    color: rgba(102, 102, 102, .5);
```
Comment (Crashillo) [inline app/assets/stylesheets/module-dashboards.scss]: Static colors?

File: app/assets/stylesheets/module-dashboards.scss [hunk 7]
```diff
@@ -237,3 +242,354 @@
...
+    text-align: right;
+    display: inline-block;
+    width: 100%;
+    text-transform: uppercase;
+  }
+
+  &-table-item-right-table {
+    border-left: 7px solid rgba(1, 120, 168, .2);
```
Comment (Crashillo) [inline app/assets/stylesheets/module-dashboards.scss]: Static colors?

File: app/assets/stylesheets/module-dashboards.scss [hunk 8]
```diff
@@ -237,3 +242,354 @@
...
+  }
+
+  &-table-item-right-table-element-bold {
+    font-weight: bold;
+  }
+
+  &-table-item-right-table-text,
+  &-table-item-right-table-amount {
+    font-size: 1rem;
+    color: #666;
```
Comment (Crashillo) [inline app/assets/stylesheets/module-dashboards.scss]: Static colors?

File: app/javascript/gobierto_dashboards/index.js [hunk 9]
```diff
@@ -26,4 +27,15 @@ document.addEventListener('DOMContentLoaded', () => {
       dataDownloadEndpoint: subsidiesAppNode.dataset.dataDownloadEndpoint,
     });
   }
+
+  //FIX-ME
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/index.js]: FIXME still on? If so, provide explanation

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    if (entryPoint) {
+      const htmlRouterBlock = `<router-view></router-view>`;
+
+      entryPoint.innerHTML = htmlRouterBlock;
+
+      const Home = () => import("../webapp/containers/costs/Home.vue");
+      const TableSecondLevel = () => import("../webapp/containers/costs/TableSecondLevel.vue");
+      const TableItem = () => import("../webapp/containers/costs/TableItem.vue");
+
+      Promise.all([getRemoteData(options.costsEndpoint)]).then((rawData) => {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: Perhaps you don't need to do a Promise.all if you have just one promise

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    //Some values are empty, so we need to transform to zero
+    function nanToZero(val) {
+       val = +val || 0
+       return val;
+    }
+
+    //Array with all the strings that we've to convert to Number
+    const amountStrings = [ 'cd_bens_i_serveis', 'cd_cost_personal', 'cost_directe_2018' , 'cost_indirecte_2018', 'cost_total_2018', 'costpers2018', 'costrestadir2018', 'cost_per_habitant', 'ingressos', 'respecte_ambit', 'taxa_o_preu_public', 'cd_serveis_exteriors', 'cd_transferencies', 'cd_equipaments', 'ingres_cost', 'subvencio']
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js:106]: Hmmmm, static texts? why so?

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    for (let cost of rawData) {
+      for (let index = 0; index < cost.length; index++) {
+        let d = cost[index]
+
+        for (let amounts = 0; amounts < amountStrings.length; amounts++) {
+          d[amountStrings[amounts]] = convertStringToNumbers(d[amountStrings[amounts]])
+        }
+
+        //Include a year column
+        if (d.cost_directe_2018) {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: This function (setGlobalVariables) are very very very specific to a site. Perhaps we should think in configuration file? cc/ @ferblape

File: app/javascript/gobierto_dashboards/webapp/containers/costs/Detail.vue [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<template>
+  <div class="pure-u-1 gutters m_b_1">
+    <Table
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/Detail.vue]: If you move the classes to the Table componet, does the markup break? If don't, you may remove the outer `div`

File: app/javascript/gobierto_dashboards/webapp/containers/costs/Distribution.vue [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    return {
+      labelDistribution: I18n.t("gobierto_dashboards.dashboards.costs.distribution") || "",
+    }
+  },
+  created() {
+    this.getDataForBubbles()
+  },
+  methods: {
+    getDataForBubbles() {
+
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/Distribution.vue:31]: It's really difficult, I'd say almost impossible, to figure out what this function does 😸

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue [hunk 15]
```diff
@@ -0,0 +1,81 @@
+<template>
+  <table class="gobierto-dashboards-table">
+    <tbody>
+      <tr
+        v-for="{ agrupacio, cost_directe_2018, cost_indirecte_2018, cost_total_2018, cost_per_habitant, ingressos, respecte_ambit, ordre_agrupacio } in groupData"
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue:5]: The bad point of this is about we are strongly linking the Costs component, to the results of an specific API (from a Catalonia place). That implies, these components aren't gonna work with any different endpoints. Is this OK? cc/ @ferblape

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  props: {
+    items: {
+      type: Array,
+      default: () => []
+    }
+  },
+  computed: {
+    groupData() {
+      //reduce to sum all values for agrupacio
+      let dataGroup = []
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue]: What if you preprocess all this data calculations before the vue-app starts? In this way you won't require to spend processing time when the component is mounted/unmounted...

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableItem.vue [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+        </router-link>
+      </div>
+    </TableHeader>
+    <table
+      v-if="items.length"
+      class="gobierto-dashboards-table gobierto-dashboards-table--subheader"
+    >
+      <tbody>
+        <tr
+          v-for="{ agrupacio, nomact, cost_directe_2018, cost_indirecte_2018, cost_total_2018, cost_per_habitant, ingressos, respecte_ambit } in dataGroup"
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableItem.vue:28]: Similar as I asked before

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableItem.vue [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+      income: ''
+    }
+  },
+  computed: {
+    //Get the total of all values in the cost table
+    totalCost() {
+      return this.equiptments + this.transferences + this.externalServices + this.goodServices + this.costPersonal
+    },
+    //Get the total of all values in the incomes table
+    totalIncomes() {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableItem.vue:263]: Good use of the computed props, young Padawan

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableSubHeader.vue [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+  created() {
+    const {
+      params: {
+        id: agrupacioId
+      }
+    } = this.$route
+    this.agrupacioData(agrupacioId)
+  },
+  methods: {
+    agrupacioData(id) {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableSubHeader.vue:54]: This is quite similar to the other calculation I suggested to move outside, isn't?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_dashboards/webapp/containers/costs/Distribution.vue (hunk 14 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_dashboards/webapp/containers/costs/Distribution.vue:31]: so sorry, bubbles chart are for the next PR

File: app/javascript/gobierto_dashboards/modules/costs_controller.js (hunk 11 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_dashboards/modules/costs_controller.js:106]: I need to convert to `Numbers` the values that come as formatted strings as "1,134,345.45". So, i've created an array with those values

Comment (Crashillo) [reply app/javascript/gobierto_dashboards/modules/costs_controller.js:106]: My point of this was that you're binding these column names to the application. As you can see, the names comes from Mataró, so you're setting this code to be full-dependant of Mataró fields. Barcelona costs? doesn't work. Zaragoza costs? doesn't work. Marbella costs? doesn't work.

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+
+    for (let index = 0; index < rawData.length; index++) {
+      let d = rawData[index]
+
+      for (let amounts = 0; amounts < amountStrings.length; amounts++) {
+        d[amountStrings[amounts]] = convertStringToNumbers(d[amountStrings[amounts]])
+      }
+
+      //Include a year column
+      if (d.cost_directe_2018) {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js:116]: Same as I mentioned previously

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue (hunk 15 above)
Comment (Crashillo) [reply app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue:5]: Same as I mentioned previously

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableSubHeader.vue [hunk 21]
```diff
@@ -0,0 +1,59 @@
+<template>
+  <table class="gobierto-dashboards-table gobierto-dashboards-table--subheader">
+    <tbody>
+      <tr
+        v-for="{ agrupacio, cost_directe_2018, cost_indirecte_2018, cost_total_2018, cost_per_habitant, ingressos, respecte_ambit } in dataGroup"
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/TableSubHeader.vue:5]: Same

**Review by Crashillo — APPROVED** [review]
The comments I made are about functional logic. If you all are OK having such strong link between the application and one particular client, go ahead.

File: app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue (hunk 15 above)
Comment (ferblape) [reply app/javascript/gobierto_dashboards/webapp/containers/costs/TableFirstLevel.vue:5]: Yes, this is fine for the time being. For the moment this is a customized dashboard for this client. It should be moved to an engine at some point but for the moment we'll keep it in Gobierto core.

**Comment by ferblape** [comment]
Closed in favor of #3145

**Pull request closed without merging**